### PR TITLE
8328828: Monkey Tester Application Part 4

### PR DESCRIPTION
**Pull request by andy-goryachev-oracle**
Further additions to the MonkeyTester application:

- tree/table view column property sheets ✔
- new pages: button, checkbox, hyperlink, menu bar, radio button, toggle button, canvas, split menu button, dnd ✔
- properties... context menu ✔
- context menu: table columns, toolbar items ✔
- control.context menu option
- property monitor tool ✔
- accessibility logging ✔

- editable combobox for double values with pre-set list (e.g. Region: min/max width etc.)
- Page -> Snapped Split Panes menu option for the control container

This PR simply syncs up with the dev repository
https://github.com/andy-goryachev-oracle/MonkeyTest

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8328828](https://bugs.openjdk.org/browse/JDK-8328828): Monkey Tester Application Part 4 (**Enhancement** - P4)

### Reviewers
 * [Ambarish Rapte](https://openjdk.org/census#arapte) (@arapte - **Reviewer**)
 * [Lukasz Kostyra](https://openjdk.org/census#lkostyra) (@lukostyra - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1560/head:pull/1560` \
`$ git checkout pull/1560`

Update a local copy of the PR: \
`$ git checkout pull/1560` \
`$ git pull https://git.openjdk.org/jfx.git pull/1560/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1560`

View PR using the GUI difftool: \
`$ git pr show -t 1560`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1560.diff">https://git.openjdk.org/jfx/pull/1560.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1560#issuecomment-2338985483)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
I did an initial test run, and will do more later. One quick thing I noticed is that it looks like the prefs file consumed by the version of `MonkeyTester` in this PR is not compatible with prefs file produced by the version in `jfx:master`.

Since saving and loading prefs is enabled by default, this causes some interesting behavior when switching between an older and newer `MonkeyTester`.

Try the following:

1. Remove `$HOME/.MonkeyTester`
2. Launch the current `MonkeyTester` from `jfx:master`
3. Navigate to a few demos, such as `AreaChart`, `BarChart`, `ChoiceBox`, etc.
4. Quit
5. Launch the `MonkeyTester` from this PR

BUG: the background of the demos previously navigated to is now a linear gradient. If you remove `$HOME/.MonkeyTester` between steps 4 and 5, the bug doesn't appear.

**Comment by andy-goryachev-oracle** [comment]
yes, this is expected.  the preferences simply remembers the index rather than anything else.

a more intelligent option would be to remember a toString or a hash(toString) of a choice.

**Comment by kevinrushforth** [comment]
Another possibility would be to have a version number associated with the prefs file, bump it when you make an incompatible change, and ignore the prefs on load if incompatible. Just a suggestion.

Reviewers: @lukostyra @arapte 

/reviewers 2

**Comment by kevinrushforth** [comment]
Btw, I'll still play around with it and add comments with any questions or observations, but I'll ask Ambarish and Lukasz to be the reviewers.

**Comment by andy-goryachev-oracle** [comment]
thank you for the suggestions!  

The workaround for now is to nuke the preferences file (I know, bad), or simply selecting the right choice in the UI.

The issue _could_ be solved by remembering `toString` (or `hash(toString)` value for privacy-conscious implementation) but there is no guarantee that we'll get an Object.toString() value instead (i.e. SomeClass@1234).  It's even worse when the application is obfuscated - there will be no guarantee that the class names remain the same from build to build.  This is surprisingly complicated issue, I have no solution yet (see https://github.com/andy-goryachev/AppFramework from which I lifted the code with the permission of, erm, author).

**Comment by arapte** [comment]
Providing a few comments observed when running MonkeyTester app.

1. Z transformations Scale Z / Translate Z can be avoided for 2D controls / shapes
2. TreeTableView height is fixed. It always shows 1 row. Changes in TreeTableView->Root does not have any effect. Height of TTV could vary with the number of rows.
3. Button: Button does not receive ENTER key press event when set as default button. 3.1 Check default button. 3.2 Move focus on button by clicking on it 3.3 Press ENTER key => button does not show pressed effect. Press space button, button shows pressed effect.
4. ChoiceBox/ColorPicker/MenuBar  background cannot be set back to the default. Changing back the background to <null> has no effect
5. No pages were getting rendered after following NPE was observed with RadioButton sample: Attaching the user preferences file with which issue can be reproduced [ui-settings.properties.zip](https://github.com/user-attachments/files/16978042/ui-settings.properties.zip)

 Exception in thread "JavaFX Application Thread" java.lang.NullPointerException: Cannot invoke "javafx.geometry.Pos.getHpos()" because the return value of "javafx.scene.control.RadioButton.getAlignment()" is null
	at javafx.controls@24-internal/javafx.scene.control.skin.RadioButtonSkin.layoutChildren(RadioButtonSkin.java:136)
	at javafx.controls@24-internal/javafx.scene.control.Control.layoutChildren(Control.java:612)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1208)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Parent.layout(Parent.java:1215)
	at javafx.graphics@24-internal/javafx.scene.Scene.doLayoutPass(Scene.java:596)
	at javafx.graphics@24-internal/javafx.scene.Scene$ScenePulseListener.pulse(Scene.java:2603)
	at javafx.graphics@24-internal/com.sun.javafx.tk.Toolkit.lambda$runPulse$2(Toolkit.java:401)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:400)
	at javafx.graphics@24-internal/com.sun.javafx.tk.Toolkit.runPulse(Toolkit.java:400)
	at javafx.graphics@24-internal/com.sun.javafx.tk.Toolkit.firePulse(Toolkit.java:430)
	at javafx.graphics@24-internal/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:598)
	at javafx.graphics@24-internal/com.sun.javafx.tk.quantum.QuantumToolkit.pulse(QuantumToolkit.java:578)
	at javafx.graphics@24-internal/com.sun.javafx.tk.quantum.QuantumToolkit.pulseFromQueue(QuantumToolkit.java:571)
	at javafx.graphics@24-internal/com.sun.javafx.tk.quantum.QuantumToolkit.lambda$runToolkit$11(QuantumToolkit.java:352)
	at javafx.graphics@24-internal/com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)

File: tests/manual/monkey/src/com/oracle/tools/fx/monkey/util/WritingSystemsDemo.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, 2024, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (andy-goryachev-oracle) [reply tests/manual/monkey/src/com/oracle/tools/fx/monkey/util/WritingSystemsDemo.java:2]: good catch, thanks!

**Comment by andy-goryachev-oracle** [comment]
@arapte thank you for testing!

> 1. Z transformations Scale Z / Translate Z can be avoided for 2D controls / shapes

I'd rather keep them, might be useful for testing (i.e. to ensure there is no change when setting these properties in 2D)

> * TreeTableView height is fixed. It always shows 1 row. Changes in TreeTableView->Root does not have any effect. Height of TTV could vary with the number of rows.

You probably need to make sure the "Root:" has a number of items, and click the root node to expand.

I did notice one thing though: setting the Background and the Fixed Cell Size 33.4 results in a weird rendering (probably because of the fractional cell size).  Not sure if it's a bug though:

Screenshot 2024-09-13 at 12 11 55


> * Button: Button does not receive ENTER key press event when set as default button.

My understanding is that a default button creates an accelerator at the Scene level, so the button rendering is unaffected when ENTER is pressed.

Try Window -> Open Modal Window and hit ENTER.  No change.

> * ChoiceBox/ColorPicker/MenuBar  background cannot be set back to the default. Changing back the background to  has no effect

I am not sure why this is happening - possibly the interaction with the skin.  Setting background back to null does not resurrect the skin's visuals, neither setting a new skin with Skin -> Set New Skin.  Only re-creating the control by switching to a different page and back does the trick.

This could be a P5 product bug, possibly.

> * No pages were getting rendered after following NPE was observed with RadioButton sample

So this is either a product bug or a case of incomplete documentation.  I've raised this issue before (see https://bugs.openjdk.org/browse/JDK-8334883), but basically some properties allow for null values and some do not.  The documentation is unclear.

In this particular case, we might have encountered a product bug: even though the code in Labeled:190 shows some attempt to deal with null property instance, it fails when the property _value_ is null:

```
    public final Pos getAlignment() { return alignment == null ? Pos.CENTER_LEFT : alignment.get(); }
```

so we need to either:
- throw NPE when setting a null value
- fix the getter to handle null value as some valid value
- consider documenting whether nulls are allowed in various javafx properties 

In any case, these are product issues and not the MT issue.  In fact, the MT serves its purpose of uncovering product issues, so thank you!

**Comment by andy-goryachev-oracle** [comment]
Just a note: removing the `io-settings.properties` file is a legitimate approach.
it is *very* difficult to create a full-proof solution that will be backward and forward compatible.

**Review by arapte — CHANGES_REQUESTED** [review]
Few quick copyright year changes required. Shall approve once changed.

**Review by arapte — APPROVED** [review]
LGTM

**Review by lukostyra — APPROVED** [review]
LGTM

**Pull request closed without merging**